File: compiler/ts/src/lexer/lexer.ts
import { Token, TokenKind, Span, KEYWORDS } from './token.js';

export class Lexer {
  private source: string;
  private pos: number;
  private line: number;
  private col: number;
  private tokens: Token[];

  constructor(source: string) {
    this.source = source;
    this.pos = 0;
    this.line = 1;
    this.col = 1;
    this.tokens = [];
  }

  tokenize(): Token[] {
    while (this.pos < this.source.length) {
      const ch = this.peek();

      // Newlines
      if (ch === '\n') {
        const start = this.pos;
        const line = this.line;
        const col = this.col;
        this.advance();
        this.tokens.push(this.makeToken(TokenKind.Newline, '\n', start, line, col));
        continue;
      }

      // Skip whitespace (spaces, tabs, carriage returns)
      if (ch === ' ' || ch === '\t' || ch === '\r') {
        this.advance();
        continue;
      }

      // Line comments: //
      if (ch === '/' && this.peekNext() === '/') {
        this.skipLineComment();
        continue;
      }

      // Block comments: /* */
      if (ch === '/' && this.peekNext() === '*') {
        this.skipBlockComment();
        continue;
      }

      // String literals
      if (ch === '"') {
        this.tokens.push(this.readString());
        continue;
      }

      // Number literals
      if (isDigit(ch)) {
        this.tokens.push(this.readNumber());
        continue;
      }

      // Identifiers and keywords
      if (isAlpha(ch) || ch === '_') {
        this.tokens.push(this.readIdentOrKeyword());
        continue;
      }

      // Operators and delimiters
      this.tokens.push(this.readOperator());
    }

    this.tokens.push(this.makeToken(TokenKind.EOF, '', this.pos, this.line, this.col));
    return this.tokens;
  }

  private advance(): string {
    const ch = this.source[this.pos];
    this.pos++;
    if (ch === '\n') {
      this.line++;
      this.col = 1;
    } else {
      this.col++;
    }
    return ch;
  }

  private peek(): string {
    if (this.pos >= this.source.length) return '\0';
    return this.source[this.pos];
  }

  private peekNext(): string {
    if (this.pos + 1 >= this.source.length) return '\0';
    return this.source[this.pos + 1];
  }

  private skipWhitespace(): void {
    while (this.pos < this.source.length) {
      const ch = this.peek();
      if (ch === ' ' || ch === '\t' || ch === '\r') {
        this.advance();
      } else {
        break;
      }
    }
  }

  private skipLineComment(): void {
    const start = this.pos;
    const line = this.line;
    const col = this.col;
    // Skip the //
    this.advance();
    this.advance();
    while (this.pos < this.source.length && this.peek() !== '\n') {
      this.advance();
    }
    const value = this.source.slice(start, this.pos);
    this.tokens.push(this.makeToken(TokenKind.Comment, value, start, line, col));
  }

  private skipBlockComment(): void {
    const start = this.pos;
    const line = this.line;
    const col = this.col;
    // Skip /*
    this.advance();
    this.advance();
    let depth = 1;
    while (this.pos < this.source.length && depth > 0) {
      if (this.peek() === '/' && this.peekNext() === '*') {
        this.advance();
        this.advance();
        depth++;
      } else if (this.peek() === '*' && this.peekNext() === '/') {
        this.advance();
        this.advance();
        depth--;
      } else {
        this.advance();
      }
    }
    const value = this.source.slice(start, this.pos);
    this.tokens.push(this.makeToken(TokenKind.Comment, value, start, line, col));
  }

  private readString(): Token {
    const start = this.pos;
    const line = this.line;
    const col = this.col;
    // Skip opening quote
    this.advance();
    let value = '"';
    while (this.pos < this.source.length && this.peek() !== '"') {
      if (this.peek() === '\\') {
        value += this.advance(); // backslash
        if (this.pos < this.source.length) {
          value += this.advance(); // escaped char
        }
      } else {
        value += this.advance();
      }
    }
    if (this.pos < this.source.length) {
      value += this.advance(); // closing quote
    }
    return this.makeToken(TokenKind.String, value, start, line, col);
  }

  private readNumber(): Token {
    const start = this.pos;
    const line = this.line;
    const col = this.col;
    let isFloat = false;

    while (this.pos < this.source.length && isDigit(this.peek())) {
      this.advance();
    }

    // Check for decimal point (but not .. operator)
    if (this.peek() === '.' && this.peekNext() !== '.' && isDigit(this.peekNext())) {
      isFloat = true;
      this.advance(); // consume '.'
      while (this.pos < this.source.length && isDigit(this.peek())) {
        this.advance();
      }
    }

    const value = this.source.slice(start, this.pos);
    const kind = isFloat ? TokenKind.Float : TokenKind.Int;
    return this.makeToken(kind, value, start, line, col);
  }

  private readIdentOrKeyword(): Token {
    const start = this.pos;
    const line = this.line;
    const col = this.col;

    while (this.pos < this.source.length && isAlphaNumOrUnderscore(this.peek())) {
      this.advance();
    }

    const value = this.source.slice(start, this.pos);

    // Check keywords map
    const kwKind = KEYWORDS.get(value);
    if (kwKind !== undefined) {
      return this.makeToken(kwKind, value, start, line, col);
    }

    // Distinguish Ident from UpperIdent by first character
    const kind = isUpper(value[0]) ? TokenKind.UpperIdent : TokenKind.Ident;
    return this.makeToken(kind, value, start, line, col);
  }

  private readOperator(): Token {
    const start = this.pos;
    const line = this.line;
    const col = this.col;
    const ch = this.advance();
    const next = this.peek();

    switch (ch) {
      case '+':
        return this.makeToken(TokenKind.Plus, '+', start, line, col);
      case '-':
        if (next === '>') {
          this.advance();
          return this.makeToken(TokenKind.Arrow, '->', start, line, col);
        }
        return this.makeToken(TokenKind.Minus, '-', start, line, col);
      case '*':
        return this.makeToken(TokenKind.Star, '*', start, line, col);
      case '/':
        return this.makeToken(TokenKind.Slash, '/', start, line, col);
      case '%':
        return this.makeToken(TokenKind.Percent, '%', start, line, col);
      case '=':
        if (next === '=') {
          this.advance();
          return this.makeToken(TokenKind.Eq, '==', start, line, col);
        }
        if (next === '>') {
          this.advance();
          return this.makeToken(TokenKind.FatArrow, '=>', start, line, col);
        }
        return this.makeToken(TokenKind.Assign, '=', start, line, col);
      case '!':
        if (next === '=') {
          this.advance();
          return this.makeToken(TokenKind.NotEq, '!=', start, line, col);
        }
        return this.makeToken(TokenKind.Not, '!', start, line, col);
      case '<':
        if (next === '=') {
          this.advance();
          return this.makeToken(TokenKind.LtEq, '<=', start, line, col);
        }
        if (next === '>') {
          this.advance();
          return this.makeToken(TokenKind.Concat, '<>', start, line, col);
        }
        return this.makeToken(TokenKind.Lt, '<', start, line, col);
      case '>':
        if (next === '=') {
          this.advance();
          return this.makeToken(TokenKind.GtEq, '>=', start, line, col);
        }
        if (next === '>') {
          this.advance();
          return this.makeToken(TokenKind.Compose, '>>', start, line, col);
        }
        return this.makeToken(TokenKind.Gt, '>', start, line, col);
      case '&':
        if (next === '&') {
          this.advance();
          return this.makeToken(TokenKind.And, '&&', start, line, col);
        }
        return this.makeToken(TokenKind.Ampersand, '&', start, line, col);
      case '|':
        if (next === '>') {
          this.advance();
          return this.makeToken(TokenKind.Pipe, '|>', start, line, col);
        }
        if (next === '|') {
          this.advance();
          return this.makeToken(TokenKind.Or, '||', start, line, col);
        }
        return this.makeToken(TokenKind.Bar, '|', start, line, col);
      case '?':
        return this.makeToken(TokenKind.Question, '?', start, line, col);
      case '.':
        if (next === '.') {
          this.advance();
          return this.makeToken(TokenKind.DotDot, '..', start, line, col);
        }
        return this.makeToken(TokenKind.Dot, '.', start, line, col);
      case ':':
        if (next === ':') {
          this.advance();
          return this.makeToken(TokenKind.ColonColon, '::', start, line, col);
        }
        return this.makeToken(TokenKind.Colon, ':', start, line, col);
      case ',':
        return this.makeToken(TokenKind.Comma, ',', start, line, col);
      case ';':
        return this.makeToken(TokenKind.Semicolon, ';', start, line, col);
      case '(':
        return this.makeToken(TokenKind.LParen, '(', start, line, col);
      case ')':
        return this.makeToken(TokenKind.RParen, ')', start, line, col);
      case '{':
        return this.makeToken(TokenKind.LBrace, '{', start, line, col);
      case '}':
        return this.makeToken(TokenKind.RBrace, '}', start, line, col);
      case '[':
        return this.makeToken(TokenKind.LBracket, '[', start, line, col);
      case ']':
        return this.makeToken(TokenKind.RBracket, ']', start, line, col);
      default:
        // Unknown character - skip it and produce an Ident token with the character
        // This is a fallback; in a real compiler we'd report an error.
        return this.makeToken(TokenKind.Ident, ch, start, line, col);
    }
  }

  private makeToken(kind: TokenKind, value: string, start: number, line: number, col: number): Token {
    return {
      kind,
      value,
      span: { start, end: this.pos, line, col },
    };
  }
}

function isDigit(ch: string): boolean {
  return ch >= '0' && ch <= '9';
}

function isAlpha(ch: string): boolean {
  return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || ch === '_';
}

function isUpper(ch: string): boolean {
  return ch >= 'A' && ch <= 'Z';
}

function isAlphaNumOrUnderscore(ch: string): boolean {
  return isAlpha(ch) || isDigit(ch);
}

File: japl-compiler/src/lexer.rs
use crate::error::{self, Span};
use crate::token::{Token, SpannedToken};

pub struct Lexer {
    input: Vec<char>,
    pos: usize,
    line: usize,
    col: usize,
    file: String,
}

impl Lexer {
    pub fn new(input: &str, file: &str) -> Self {
        Lexer {
            input: input.chars().collect(),
            pos: 0,
            line: 1,
            col: 1,
            file: file.to_string(),
        }
    }

    fn peek(&self) -> Option<char> {
        self.input.get(self.pos).copied()
    }

    fn peek2(&self) -> Option<char> {
        self.input.get(self.pos + 1).copied()
    }

    fn advance(&mut self) -> Option<char> {
        let ch = self.input.get(self.pos).copied()?;
        self.pos += 1;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        Some(ch)
    }

    fn span(&self) -> Span {
        Span { line: self.line, col: self.col }
    }

    fn skip_whitespace(&mut self) {
        while let Some(ch) = self.peek() {
            if ch.is_whitespace() {
                self.advance();
            } else if ch == '/' && self.peek2() == Some('/') {
                // Check for doc comment (///)
                if self.input.get(self.pos + 2) == Some(&'/') {
                    break; // Don't skip doc comments - let tokenize handle them
                }
                // Regular line comment
                while let Some(c) = self.peek() {
                    if c == '\n' { break; }
                    self.advance();
                }
            } else {
                break;
            }
        }
    }

    fn read_string(&mut self) -> error::Result<String> {
        let span = self.span();
        self.advance(); // skip opening quote
        let mut s = String::new();
        loop {
            match self.advance() {
                Some('"') => return Ok(s),
                Some('\\') => {
                    match self.advance() {
                        Some('n') => s.push('\n'),
                        Some('t') => s.push('\t'),
                        Some('\\') => s.push('\\'),
                        Some('"') => s.push('"'),
                        _ => return error::err("invalid escape", &self.file, Some(span)),
                    }
                }
                Some(ch) => s.push(ch),
                None => return error::err("unterminated string", &self.file, Some(span)),
            }
        }
    }

    fn read_number(&mut self) -> i64 {
        let mut n: i64 = 0;
        while let Some(ch) = self.peek() {
            if ch.is_ascii_digit() {
                n = n * 10 + (ch as i64 - '0' as i64);
                self.advance();
            } else {
                break;
            }
        }
        n
    }

    fn read_ident(&mut self) -> String {
        let mut s = String::new();
        while let Some(ch) = self.peek() {
            if ch.is_alphanumeric() || ch == '_' {
                s.push(ch);
                self.advance();
            } else {
                break;
            }
        }
        s
    }

    pub fn tokenize(&mut self) -> error::Result<Vec<SpannedToken>> {
        let mut tokens = Vec::new();
        loop {
            self.skip_whitespace();
            let span = self.span();
            match self.peek() {
                None => {
                    tokens.push(SpannedToken { token: Token::Eof, span });
                    return Ok(tokens);
                }
                Some(ch) => {
                    let token = match ch {
                        '"' => {
                            let s = self.read_string()?;
                            Token::StringLit(s)
                        }
                        '(' => { self.advance(); Token::LParen }
                        ')' => { self.advance(); Token::RParen }
                        '{' => { self.advance(); Token::LBrace }
                        '}' => { self.advance(); Token::RBrace }
                        ':' => { self.advance(); Token::Colon }
                        ',' => { self.advance(); Token::Comma }
                        '.' => { self.advance(); Token::Dot }
                        '+' => { self.advance(); Token::Plus }
                        '*' => { self.advance(); Token::Star }
                        '/' => {
                            self.advance();
                            // Check for doc comment: ///
                            if self.peek() == Some('/') && self.peek2() == Some('/') {
                                self.advance(); // second /
                                self.advance(); // third /
                                // Skip optional leading space
                                if self.peek() == Some(' ') {
                                    self.advance();
                                }
                                let mut doc = String::new();
                                while let Some(c) = self.peek() {
                                    if c == '\n' { break; }
                                    doc.push(c);
                                    self.advance();
                                }
                                Token::DocComment(doc)
                            } else {
                                Token::Slash
                            }
                        }
                        '%' => { self.advance(); Token::Percent }
                        '-' => {
                            self.advance();
                            if self.peek() == Some('>') {
                                self.advance();
                                Token::Arrow
                            } else {
                                Token::Minus
                            }
                        }
                        '=' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                                Token::EqEq
                            } else if self.peek() == Some('>') {
                                self.advance();
                                Token::FatArrow
                            } else {
                                Token::Eq
                            }
                        }
                        '!' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                                Token::BangEq
                            } else {
                                return error::err("unexpected '!'", &self.file, Some(span));
                            }
                        }
                        '<' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                                Token::LtEq
                            } else if self.peek() == Some('>') {
                                self.advance();
                                Token::Concat
                            } else {
                                Token::Lt
                            }
                        }
                        '>' => {
                            self.advance();
                            if self.peek() == Some('=') {
                                self.advance();
                                Token::GtEq
                            } else {
                                Token::Gt
                            }
                        }
                        '|' => {
                            self.advance();
                            if self.peek() == Some('>') {
                                self.advance();
                                Token::PipeOp
                            } else {
                                Token::Pipe
                            }
                        }
                        _ if ch.is_ascii_digit() => {
                            let n = self.read_number();
                            Token::Int(n)
                        }
                        _ if ch.is_alphabetic() || ch == '_' => {
                            let ident = self.read_ident();
                            match ident.as_str() {
                                "fn" => Token::Fn,
                                "let" => Token::Let,
                                "if" => Token::If,
                                "else" => Token::Else,
                                "match" => Token::Match,
                                "type" => Token::Type,
                                "foreign" => Token::Foreign,
                                "True" => Token::True,
                                "False" => Token::False,
                                "receive" => Token::Receive,
                                "import" => Token::Import,
                                "pub" => Token::Pub,
                                "const" => Token::Const,
                                "trait" => Token::Trait,
                                "opaque" => Token::Opaque,
                                "use" => Token::Use,
                                _ => Token::Ident(ident),
                            }
                        }
                        _ => {
                            return error::err(
                                format!("unexpected character '{}'", ch),
                                &self.file,
                                Some(span),
                            );
                        }
                    };
                    tokens.push(SpannedToken { token, span });
                }
            }
        }
    }
}
